data_model/src/predicate.rs

```diff
@@ -2,54 +2,155 @@
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+pub struct NonEmpty<T> {
+    head: Box<T>,
+    tail: Vec<T>,
+}
+
+macro_rules! nonempty {
+    ($h:expr, $( $t:expr ),*) => {{
+        NonEmpty {
+            head: Box::new($h),
+            tail: vec![$($t),*],
+        }
+    }};
+}
+
 /// Predicate combinator enum.
 #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub enum PredicateBox {
-    /// Logically `&&` the results of applying the two predicates.
-    And(Vec<PredicateBox>),
-    /// Logically `||` the results of applying the two predicates.
-    Or(Vec<PredicateBox>),
+// Ideally we would enforce `P: PredicateTrait<Input>` here, but I
+// couldn't find a way to do it without polluting everything
+// downstream with explicit lifetimes, since we would need to
+// store PhantomData<Input> here, and `Input`s are most often
+// references (e.g. &Value).
+pub enum GenericPredicateBox<P> {
+    /// Logically `&&` the results of applying the predicates.
+    And(NonEmpty<Self>),
+    /// Logically `||` the results of applying the predicates.
+    Or(NonEmpty<Self>),
     /// Negate the result of applying the predicate.
-    Not(Box<PredicateBox>),
+    Not(Box<Self>),
     /// The raw predicate that must be applied.
-    Raw(value::Predicate),
+    Raw(P),
 }
 
-impl PredicateBox {
+impl<P> Display for GenericPredicateBox<P>
+where
+    P: Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GenericPredicateBox::And(predicates) => {
+                write!(f, "OR(")?;
+                predicates.head.fmt(f)?;
+                for predicate in &predicates.tail {
+                    predicate.fmt(f)?;
+                }
+                write!(f, ")")
+            }
+            GenericPredicateBox::Or(predicates) => {
+                write!(f, "AND(")?;
+                predicates.head.fmt(f)?;
+                for predicate in &predicates.tail {
+                    predicate.fmt(f)?;
+                }
+                write!(f, ")")
+            }
+            GenericPredicateBox::Not(predicate) => write!(f, "NOT({})", predicate),
+            GenericPredicateBox::Raw(predicate) => predicate.fmt(f),
+        }
+    }
+}
+
+impl<P> GenericPredicateBox<P> {
     /// Construct [`PredicateBox::Raw`] variant.
     #[inline]
-    pub fn new(pred: impl Into<value::Predicate>) -> Self {
+    pub fn new<Input>(pred: impl Into<P>) -> Self
+    where
+        P: PredicateTrait<Input>,
+        Input: Copy,
+    {
         Self::Raw(pred.into())
     }
 
     /// Construct [`PredicateBox::And`] variant.
     #[inline]
-    pub fn and(left: impl Into<PredicateBox>, right: impl Into<PredicateBox>) -> Self {
-        Self::And(vec![left.into(), right.into()])
+    pub fn and(left: impl Into<Self>, right: impl Into<Self>) -> Self {
+        Self::And(nonempty![left.into(), right.into()])
```

Comment:
Why do we need `NonEmpty` if `And` and `Or` is binary operations?

Reply:
`And` and `Or` allow any number of operands, without NonEmpty it would be possible to manually construct empty `And` variant, which we can't evaluate since `PredicateSymbol` doesn't require `Default` implementation (and I think shouldn't, since it isn't really well-defined for e.g. ValidatorVerdict, and for boolean we would want different defaults in different contexts)

Reply:
> `And` and `Or` allow any number of operands

But `and/or` accepts only two arguments, so we always constructing tree-like structure `add(a, add(b, c))`.

Is user supposed to manually create multi-argument `And/Or`? 

My suggestion is that we can inspect `left` and if it is `And` for `fn and` (`Or` for `fn or`) we can append `right` to argument list.

Otherwise we can restrict `And` as `And(Self, Self)`.

Reply:
> But and/or accepts only two arguments, so we always constructing tree-like structure add(a, add(b, c)).

That is true, I was worried about user manually constructing invalid PredicateBox
> My suggestion is that we can inspect left and if it is And for fn and (Or for fn or) we can append right to argument list.

Agreed, I'll implement it